### PR TITLE
[RFC][move] Add capability constraints for reconfiguration.

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation
This change defines who has the capability to modify on-chain configs. There are two main design choices here:

1. Should the checks be enforced at caller site (`LibraVMConfig`, `LibraSystem`, etc) or at the callee site(`LibraConfig`).
    - This implementation enforces at caller site. The reason for enforcing at caller site is that we can have a batch control of who has the right to modify the config.
    - An alternate design is we can have each of the caller defines its own access control policy. This may allow finer grained control of the accounts that could modify the config

2.  How to enable non-association accounts modify the config? The major use case of this scenario is ValidatorSet's key rotation operation.
    - In this PR, @tzakian and I came up the `DelegatedLibraConfigCapability` that allows caller to temporary give out the right to modify on chain configs. 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### CODE REVIEW

File: language/stdlib/modules/libra_configs.move [hunk 1]
```diff
@@ -31,38 +40,75 @@ module LibraConfig {
         });
     }
 
-    // We can't return reference here, this returns a copy of the config value.
+    public fun apply_for_default_privilege() {
+        if(!Association::addr_is_association(Transaction::sender())) {
+            Association::apply_for_association();
+        };
```
Comment (tzakian) [inline language/stdlib/modules/libra_configs.move:46]: This check seems superfluous with the other check that was added in the `Association` module. What do you think about removing that one and keeping this?

File: language/stdlib/modules/libra_configs.move [hunk 2]
```diff
@@ -31,38 +40,75 @@ module LibraConfig {
...
     }
 
-    // Publish a new config item to a new value and trigger a reconfiguration.
-    public fun publish_new_config<Config: copyable>(payload: Config) {
-        // TODO: impose proper permission checks
-        // Callable by the any address for now.
+    // Publish a new config item.
+    public fun publish_new_config<Config: copyable>(payload: Config): Self::DelegatedLibraConfigPrivilege<Config> {
+        Transaction::assert(
+            Association::has_privilege<Self::LibraConfigPrivilege>(Transaction::sender()),
```
Comment (tzakian) [inline language/stdlib/modules/libra_configs.move:88]: `Self::` isn't needed within the module

File: language/stdlib/modules/libra_configs.move [hunk 3]
```diff
@@ -31,38 +40,75 @@ module LibraConfig {
...
         *&borrow_global<T<Config>>(addr).payload
     }
 
     // Set a config item to a new value and trigger a reconfiguration.
-    public fun set<Config: copyable>(addr: address, payload: Config) acquires T, Configuration {
-        // TODO: impose proper permission checks
-        // Callable by the any address for now.
+    public fun set<Config: copyable>(payload: Config) acquires T, Configuration {
+        Transaction::assert(
+            Association::has_privilege<Self::LibraConfigPrivilege>(Transaction::sender()),
```
Comment (tzakian) [inline language/stdlib/modules/libra_configs.move:64]: `Self::` isn't needed within the module

File: language/stdlib/modules/libra_configs.move [hunk 4]
```diff
@@ -31,38 +40,75 @@ module LibraConfig {
...
+            1
+        );
 
         move_to_sender(T{ payload });
         // We don't trigger reconfiguration here, instead we'll wait for all validators update the binary
         // to register this config into ON_CHAIN_CONFIG_REGISTRY then send another transaction to change
         // the value which triggers the reconfiguration.
+
+        return DelegatedLibraConfigPrivilege<Config> {
```
Comment (tzakian) [inline language/stdlib/modules/libra_configs.move:97]: `return` isn't needed.

File: language/stdlib/modules/libra_configs.move [hunk 5]
```diff
@@ -31,38 +40,75 @@ module LibraConfig {
...
+            delegated_address: Transaction::sender(),
+        }
+    }
+
+    public fun drop_capability<Config>(cap: DelegatedLibraConfigPrivilege<Config>) {
+        let DelegatedLibraConfigPrivilege<Config> { delegated_address: _ } = cap;
```
Comment (tzakian) [inline language/stdlib/modules/libra_configs.move:103]: No need to for the `let` here :) 
```suggestion
        DelegatedLibraConfigPrivilege<Config> { delegated_address: _ } = cap;
```

File: language/stdlib/modules/libra_version.move [hunk 6]
```diff
@@ -9,10 +9,13 @@ module LibraVersion {
     }
 
     public fun initialize() {
+        Transaction::assert(Transaction::sender() == LibraConfig::default_config_address(), 1);
 
-        LibraConfig::publish_new_config<Self::T>(T {
+        let cap = LibraConfig::publish_new_config<Self::T>(T {
```
Comment (tzakian) [inline language/stdlib/modules/libra_version.move:14]: You shouldn't need to instantiate `publish_new_config` or `drop_capability` below; the compiler should be able to infer this information for you :)

File: language/stdlib/modules/libra_version.move [hunk 7]
```diff
@@ -23,6 +26,6 @@ module LibraVersion {
             25
         );
 
-        LibraConfig::set<Self::T>(Transaction::sender(), T { major } )
+        LibraConfig::set<Self::T>( T { major } )
```
Comment (tzakian) [inline language/stdlib/modules/libra_version.move:29]: Same here; shouldn't need the instantiation

File: language/stdlib/modules/libra_vm_config.move [hunk 8]
```diff
@@ -34,18 +34,19 @@ module LibraVMConfig {
         instruction_schedule: vector<u8>,
         native_schedule: vector<u8>
     ) {
-        LibraConfig::publish_new_config<Self::T>(T {
+        let cap = LibraConfig::publish_new_config<Self::T>(T {
```
Comment (tzakian) [inline language/stdlib/modules/libra_vm_config.move:37]: Can remove this instantiation.

File: language/stdlib/modules/libra_vm_config.move [hunk 9]
```diff
@@ -34,18 +34,19 @@ module LibraVMConfig {
...
             publishing_option,
             gas_schedule: GasSchedule {
                 instruction_schedule,
                 native_schedule,
             }
-        })
+        });
+        LibraConfig::drop_capability<Self::T>(cap);
```
Comment (tzakian) [inline language/stdlib/modules/libra_vm_config.move:44]: Same here.

File: language/stdlib/modules/libra_vm_config.move [hunk 10]
```diff
@@ -34,18 +34,19 @@ module LibraVMConfig {
...
     }
 
     public fun set_publishing_option(publishing_option: vector<u8>) {
         let current_config = LibraConfig::get<Self::T>(Transaction::sender());
         current_config.publishing_option = publishing_option;
-        LibraConfig::set<Self::T>(Transaction::sender(), current_config )
+        LibraConfig::set<Self::T>( current_config )
```
Comment (tzakian) [inline language/stdlib/modules/libra_vm_config.move:50]: Can remove this instantiation.

**Comment by sblackshear** [comment]
Couple of thoughts:

> An alternate design is we can have each of the caller defines its own access control policy. This may allow finer grained control of the accounts that could modify the config

I think we will indeed need something more fine-grained here; one capability per config seems like the most straightforward option. The sensitivity, frequency of updates, and additional (possibly one-off) constraints will vary significantly across configs.

I think the validator key rotation is already an example of the third point. Imagine if  we started off without `DelegatedLibraConfigPrivilege`, then a new config came along that needed a feature like this. It would be much easier to add this feature in the module that defines the new config than it would be to define a new general-purpose capability in `LibraConfig`.

One idea:
- Each simple config implementation does something like
```
public fun set(major: u64) {
  assert(Association::has_privilege<PrivilegedCapability<T>>(sender()));
  // setter
}
```

, where `T` is defined in the config module.

- Fancier configs can use fancier `T`'s/fancier assertions. For example, the validator set key rotation case could be something like:

```
resource struct T {
  delegate: address
}

// let the owner of PrivilegedCapability of T set the delegate

public fun rotate_key(new_key: bytearray) {
  assert(Association::exists<PrivilegedCapability<T>>); // as normal
  //
}

public fun rotate_key_as_delegate(new_key: bytearray, owner: address) {
  let cap = PrivilegedCapability<T>>(owner);
  assert(delegate(&cap.t) == sender());
}
```

Under this approach:
- each config capability decides whether it can be delegated
- each config capability can decide on additional constraints (e.g., rate limiting for validator  key rotation)
- each config capability can be given to a single address (if that is indeed what we want)
- config capabilities can be spread among different addresses depending on their sensitivity or frequency of update
- we can dynamically switch the address -> capability allocation over time as requirements change

File: language/stdlib/modules/association.move [hunk 11]
```diff
@@ -38,6 +38,9 @@ module Association {
 
     // Publish a specific privilege under the sending account.
     public fun apply_for_privilege<Privilege>() {
+        if(::exists<PrivilegedCapability<Privilege>>(Transaction::sender())) {
+            return
+        };
```
[the remark this answers is not in the document]
Comment (bob-wilson) [reply language/stdlib/modules/association.move:43]: Shouldn't there also be a space after the "if"?

**Comment by runtian-zhou** [comment]
> I think the validator key rotation is already an example of the third point. Imagine if we started off without `DelegatedLibraConfigPrivilege`

Let me see if I understand the idea correctly. Do you mean we define `PrivilegedCapability` in `LibraConfig` module? If that's the case I would argue that we do actually need `DelegatedLibraConfigPrivilege`. The main issue is that we actually need some access control on the `LibraConfig::set`, the reason is that it will emit reconfiguration event, thus it can't be callable from arbitrary address.

Meanwhile I can try to come up with an alternate design, imagine we have `PrivilegedCapability<T>` defined in `LibraConfig`, but it will not be an association privilege but just a resource. The apis in `LibraConfig` will look like following:

```
pub fun set<Config>(cap: &PrivilegedCapability<Config>, config: Config);
pub fun publish_new_config<Config>(config: Config): PrivilegedCapability<Config>;
```

One problem to this design is that `publish_new_config` should be called by address where the config is going to be stored, since we only have `move_to_sender`. However the ability to add new config item should strictly be restricted, which implies it should be some sorts of association capability. One potential solution is that we have an association privilege to add a new config and a resource based capability to modify existing configs.

**Comment by zekun000** [comment]
let me put my preliminary thoughts here, briefly discussed with @runtian-zhou before, would like to hear opinions @sblackshear @tzakian 

1. Each config defines its own Privilege for who can update what e.g. ValidatorSetPrivilege would be granted for validators to rotate their keys, but they don't need DelegatedLibraConfigPrivilege to set the config value.
2. `set::<T>` doesn't do permission check but it's guarded implicitly at callers given the constructor of each config by its own Privileges no one can mess up with random set::<T> in LibraConfig.
3. LibraConfigPrivilege is only granted to association for the function `publish_new_config`, 
`publish_new_config` is responsible to guarantee only well-behaved config modules are published, in other words, this privilege guards the constructor of libra config

**Comment by zekun000** [comment]
>>> One problem to this design is that publish_new_config should be called by address where the config is going to be stored, since we only have move_to_sender

I didn't realize this constraint, how does mint work? isn't it creating a new resource for the account different than sender?

**Comment by tzakian** [comment]
I'm still processing the different comments at the moment. But I wanted to give some of my thoughts here as well. 

The way I view this as working is that:
* We have _a default permission_: when a config of type `Module::T` is published, the publishing address has the ability to control the `Module::T` config resource at that address always.
* We have a module/operations permission: the `DelegatedLibraConfigCapability<Module::T>` allows the declaring module to implement finer-grained permissioning around updates to the config if it desires using this. The module could then implement finer-grained permissions checking around the borrowing and passing of this.

To see an example of the `DelegatedLibraConfigCapability` and how it works with permissioning within the module that declares the `Config` type let's take a look at rate-limited rotation:

```rust
module MyAwesomeModule {
    use 0x0::LibraTimeStamp;
	resource struct RateLimit {
         last_rotate: u64,
         rate_limit: u64,
    }
    resource struct MyConfig {
      // cool data here!
    }
    resource struct ConfigCap {
    	cap: LibraConfig::DelegatedLibraConfigPrivilege<MyConfig>,
    }
    public fun initialize() {
        assert(sender() == 0xCAFE, 0);
    	let cap = LibraConfig::publish_new_config(MyConfig{});
		// Give me the ability to update the config from this module
		move_to_sender(ConfigCap { cap });
		// Add in any resources needed for enforcing finer-grained access controls
        move_to_sender(RateLimit{
			last_rotate: LibraTimeStamp::now_microseconds(),
  			rate_limit: 42,
   		})
    }
    
   	public fun rate_limited_rotate(update: MyConfig) acquires RateLimit {
   		let current_time = LibraTimeStamp::now_microseconds();
		let limit = borrow_global_mut<RateLimit>(0xCAFE);
		// Do my fine-grained local checks
		assert(current_time >= limit.last_rotate + limit.rate_limit, 1);
		limit.last_rotate = current_time;
		// Satisfied that it all checks out, allow the operation to take place
		LibraConfig::set_with_capability(
			&borrow_global<ConfigCap>(0xCAFE).cap,
			update
		);
	}
}

```

**Comment by sblackshear** [comment]
Taking a step back: am I right in understanding that this is what we want from configs?

1. Each config should be globally unique
2. Only an Association account has the privilege to add new configs
3. Only an Association has the privilege to update existing configs
4. Each config should emit a `NewEpochEvent` when updated
5. Only a config update should emit a `NewEpochEvent`


And if so, is it the case that:

- (1) is not currently enforced (i.e., anyone with a `LibraConfigPrivilege` can publish a copy of the config under their address)
- The `LibraConfigs::set` function is designed to ensure (2) + (3) via the `LibraConfigPrivilege` assert
- The `LibraConfigs::set` function is designed to ensure (4) via the call to `reconfigure_`
- (5) is somewhat weakly enforced because `reconfigure` can also be called from outside by anyone with a `LibraConfigPrivilege`

**Comment by davidiw** [comment]
What you say makes sense to me Sam with some caveat:
* A modification to a config may not require a reconfig -- for example updating a set of fullnode endpoints is purely informational for the external world and does not impact anything internal to the system -- this should be a config but it shouldn't require a reconfiguration
* Why limit who can create configs? Is this because of the reconfiguration ability? It seems potentially desirable to separate the two?
* Why do we care about globally uniqueness?

**Comment by sblackshear** [comment]
> A modification to a config may not require a reconfig -- for example updating a set of fullnode endpoints is purely informational for the external world and does not impact anything internal to the system -- this should be a config but it shouldn't require a reconfiguration

Good point. I'm not sure how/if we can represent such a config today. I believe we just reconfigure on everything to be conservative (since config changes are privileged and reconfiguration is fairly cheap).

> Why limit who can create configs? Is this because of the reconfiguration ability?

Yes, I think that is the only reason.

> Why do we care about globally uniqueness?

It's important for clients/Move code to know which config is the "real" one and global uniqueness seems like safest way to achieve that--for example, it would be confusing if there were two different copies of the gas cost table.

**Comment by zekun000** [comment]
> A modification to a config may not require a reconfig -- for example updating a set of fullnode endpoints is purely informational for the external world and does not impact anything internal to the system -- this should be a config but it shouldn't require a reconfiguration

This was discussed before, it's mainly simple to implement, if we have time we can further separate them.

> 3. Only an Association has the privilege to update existing configs

I thought this can be relaxed to what each config module defines without delegating as long as only association can add new config modules

**Comment by runtian-zhou** [comment]
> Taking a step back: am I right in understanding that this is what we want from configs?
> 
> 1. Each config should be globally unique

Not necessarily. Ultimately it's the rust code that determines which config under some address is being loaded.
 
> 2. Only an Association account has the privilege to add new configs

I do believe so.
> 3. Only an Association has the privilege to update existing configs

This should hold true in general. Rotating authentication key should be one exception tho, as  it will be each individual validators sending such transaction.

> 4. Each config should emit a `NewEpochEvent` when updated

Yes for now. It should be fairly easy to add a boolean flag so that we don't trigger a `NewEpochEvent` when some certain flags are updated.

> 5. Only a config update should emit a `NewEpochEvent`
 
Yes

> And if so, is it the case that:
> 
> * (1) is not currently enforced (i.e., anyone with a `LibraConfigPrivilege` can publish a copy of the config under their address)

Yes, and this is intentional as we might want to split different configs under different accounts so that we can specify different signature rules for modifying them.

> * The `LibraConfigs::set` function is designed to ensure (2) + (3) via the `LibraConfigPrivilege` assert
> * The `LibraConfigs::set` function is designed to ensure (4) via the call to `reconfigure_`

Correct! To be specific, `DelegatedLibraConfigPrivilege` is the backdoor for us to handle the key rotation scenario.

> * (5) is somewhat weakly enforced because `reconfigure` can also be called from outside by anyone with a `LibraConfigPrivilege`

That's true. We might also want to make this function private. The only reason I left it there is because we had some test cases of this and I'm not sure if there's any real life use cases over there. 

> Why limit who can create configs? Is this because of the reconfiguration ability?

That's a very good point! We actually don't do reconfiguration at config creation time. So we might as well get rid of such constraint. The only problem tho is that it can make account structure look messy, as arbitrary account can have some garbage data that can never be used. (And the even worse case is that it kind of allow accounts to store arbitrary data on chain once we go to open script mode as the config can act as a container for data that shouldn't supposed to be published).

**Comment by runtian-zhou** [comment]
After a discussion with @sblackshear, @tzakian , @valerini , @zekun000 , we came up with this simpler model:

- All Config item will be stored under one address. This will give us the global uniqueness property.
- Creating a new config item will require an association privilege and returns a `SetConfigCapability`. The code will look like following:
```
LibraConfig::create_config<Config>(config: Config, needs_reconfig: bool): SetConfigCapability<Config> {
    Association::has_privilege<CreateConfigPrivilege>(Transaction::sender());
    ...
}
```

By calling into this function, the association is granting the module of Config type the capability to update this Config item, and it will be association's job to make sure if sufficient access control policy is enforced within the module. This also answers @davidiw 's question of why we would like to limit who can publish a config item.

- Each config implementation module will need to store `SetConfigCapability<Config>` under the config address so that each module can use such resource as the access policy controller.

To summarize with answer to Sam's previous question:
1. Config will be globally unique and stored under one address.
2. Only an Association account has the privilege to add new configs.
3. There's no implicit check for setting up a new config value. Each config modules will need to specify their own access control policy in their implementation to make sure it cannot be abused.
4. Each config module can decide if reconfiguration is required when it's modified.

**Comment by valerini** [comment]
Quick question: for ValidatorSet, we want validators (or their delegates) to be able to call an "update" (that will copy ValidatorConfigs from validators' accounts into the ValidatorSet config), but we do not want validators to be able to add/remove validators. Will then these be two different privileges: one for updating the configs and one for adding/removing configs? Or will there be one privilege per individual ValidatorConfig?

**Comment by runtian-zhou** [comment]
> Quick question: for ValidatorSet, we want validators (or their delegates) to be able to call an "update" (that will copy ValidatorConfigs from validators' accounts into the ValidatorSet config), but we do not want validators to be able to add/remove validators. Will then these be two different privileges: one for updating the configs and one for adding/removing configs? Or will there be one privilege per individual ValidatorConfig?

The LibraConfig module will grant the LibraSystem the ability to change config values to anything. It's LibraSystem's role to make sure the extra access control policy that we might want to enforce. In this case, we would like the LibraSystem module to check validators are only allowed to change their keys.

File: language/stdlib/modules/libra_version.move [hunk 12]
```diff
@@ -1,28 +1,45 @@
...
+        });
+
+        move_to_sender(CapabilityHolder{ cap });
     }
 
-    public fun set(major: u64) {
-        let old_config = LibraConfig::get<Self::T>(Transaction::sender());
+    public fun set(major: u64) acquires CapabilityHolder{
+        Transaction::assert(
+            Transaction::sender() == LibraConfig::default_config_address() ||
```
Comment (sblackshear) [inline language/stdlib/modules/libra_version.move:28]: Why can the config address call this?

File: language/stdlib/modules/libra_configs.move [hunk 13]
```diff
@@ -31,38 +37,58 @@ module LibraConfig {
...
         Transaction::assert(::exists<T<Config>>(addr), 24);
         *&borrow_global<T<Config>>(addr).payload
     }
 
     // Set a config item to a new value and trigger a reconfiguration.
-    public fun set<Config: copyable>(addr: address, payload: Config) acquires T, Configuration {
-        // TODO: impose proper permission checks
-        // Callable by the any address for now.
-
+    public fun set<Config: copyable>(_cap: &ConfigModifierPrivilege<Config>, payload: Config) acquires T, Configuration {
```
Comment (sblackshear) [inline language/stdlib/modules/libra_configs.move:58]: If we changed this API to `public fun set<Config: copyable>(payload: Config)` and checked `exists<ConfigModifierCapability<Config>>(sender)` in the body, could we get rid of this unfortunate `CapabilityHolder` pattern?

Even if we need that pattern in some places (I suspect we do for validator key rotation), if we could provide two APIs for `set` (one that checks `exist` and one that takes the capability as a parameter) that gets rid of most uses of the pattern, I think that would be a nice simplification.

File: language/stdlib/modules/libra_configs.move [hunk 14]
```diff
@@ -31,38 +37,58 @@ module LibraConfig {
...
 
         move_to_sender(T{ payload });
         // We don't trigger reconfiguration here, instead we'll wait for all validators update the binary
         // to register this config into ON_CHAIN_CONFIG_REGISTRY then send another transaction to change
         // the value which triggers the reconfiguration.
+
+        return ConfigModifierPrivilege<Config> {}
+    }
+
+    public fun drop_capability<Config>(cap: ConfigModifierPrivilege<Config>) {
```
Comment (sblackshear) [inline language/stdlib/modules/libra_configs.move:82]: I think this is now unused--can we kill it if so?

File: language/stdlib/modules/libra_configs.move [hunk 15]
```diff
@@ -17,12 +18,17 @@ module LibraConfig {
         events: Event::EventHandle<NewEpochEvent>,
     }
 
-    // This can only be invoked by the Association address, and only a single time.
+    // Accounts with this association privilege can publish new config under default_address
+    struct ConfigCreatorPrivilege {}
```
[the remark this answers is not in the document]
Comment (runtian-zhou) [reply language/stdlib/modules/libra_configs.move:22]: I love `CreateConfigCapability` & `ModifyConfigCapability` !

File: language/stdlib/modules/libra_configs.move [hunk 16]
```diff
@@ -31,38 +37,58 @@ module LibraConfig {
...
         reconfigure_();
     }
 
-    // Publish a new config item to a new value and trigger a reconfiguration.
-    public fun publish_new_config<Config: copyable>(payload: Config) {
-        // TODO: impose proper permission checks
-        // Callable by the any address for now.
+    // Publish a new config item.
+    public fun publish_new_config<Config: copyable>(payload: Config): ConfigModifierPrivilege<Config> {
+        Transaction::assert(
```
[the remark this answers is not in the document]
Comment (runtian-zhou) [reply language/stdlib/modules/libra_configs.move:69]: Do you mean `publish<T>(payload);  set<T>(cap, T)`? `get` should always succeed because get is always reading the config data from the config address.

File: language/stdlib/modules/libra_version.move (hunk 12 above)
Comment (runtian-zhou) [reply language/stdlib/modules/libra_version.move:28]: I think that depends on how we are planning to control who can set this flag up. IMHO we want to set up one account uniquely for each of the config so that in the future we can just replace this assert to work with one special address instead of the two well known address.

File: language/stdlib/modules/libra_configs.move (hunk 13 above)
Comment (runtian-zhou) [reply language/stdlib/modules/libra_configs.move:58]: You are indeed right about this. This pattern will fail when we try to rotate validator key. I do like the two api suggestion. I would propose the following api:
Either:
``` 
fun publish_new_config<Config>(t: Config): ModifyConfigCapability<Config>;
fun store_config_capability<Config>(cap: ModifyConfigCapability<Config>);
fun set<Config>(t: Config);
fun set_with_capability<Config>(cap: & ModifyConfigCapability<Config>, t: Config);
```
Or:
```
fun publish_new_config_with_default_capability<Config>(t: Config);
fun publish_new_config<Config>(t: Config): ModifyConfigCapability<Config>;
fun set<Config>(t: Config);
fun set_with_capability<Config>(cap: & ModifyConfigCapability<Config>, t: Config);
```

The problem to both approaches is that we loose the ability to delegate the address in the default mode. To modify configs published in such way, the transaction sender has to be config address (0xF1A95). We might want to have a default delegator so that one other address will be able to modify such value as well. So it will look something like:
 ```
fun publish_new_config_with_default_capability<Config>(t: Config, owner_address: Address);
// Only owner_address can call this function 
fun set<Config>(t: Config);
```

File: language/stdlib/modules/libra_configs.move (hunk 16 above)
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/libra_configs.move:69]: > Do you mean publish<T>(payload); set<T>(cap, T)

Yes, my bad.

> get should always succeed because get is always reading the config data from the config address.

But what if an address other than 0xF1A95 calls `publish_new_config`?

File: language/stdlib/modules/libra_configs.move (hunk 13 above)
Comment (sblackshear) [reply language/stdlib/modules/libra_configs.move:58]: I like the first API you suggest.

> The problem to both approaches is that we loose the ability to delegate the address in the default mode

Yeah, this is a tricky problem... But I think we can solve it with your first API + a small extension using the `Offer` module:

```
publicfun publish_new_config_with_delegate<Config>(t: Config, delegate: address)  {
   // ...
   Offer::create(modify_config_cap, delegate) // offer the modify_cap to delegate
}

public fun claim_delegated_modify_config(offer_address: address) {
  move_to_sender(Offer::redeem(offer_addres))
}
```

File: language/stdlib/modules/libra_configs.move (hunk 16 above)
[the remark this answers is not in the document]
Comment (runtian-zhou) [reply language/stdlib/modules/libra_configs.move:98]: Do you think we should allow that? In my previous mindset I thought accounts with such capability should be able to publish new config. This allows different configs to resides in different addresses, which can be good or bad.

File: language/stdlib/modules/libra_configs.move (hunk 13 above)
Comment (runtian-zhou) [reply language/stdlib/modules/libra_configs.move]: I think we ended up very similar path! How do you feel about the current api?

File: language/stdlib/modules/libra_configs.move (hunk 16 above)
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/libra_configs.move:98]: I don't think we should allow it, but without the explicit address check it is allowed :)

File: language/stdlib/modules/libra_configs.move (hunk 13 above)
Comment (sblackshear) [reply language/stdlib/modules/libra_configs.move]: Getting there! I think that having both `DefaultModifyConfigCapability` and `ModifyConfigCapability` is a bit confusing.

I would prefer to have the simple model where `ModifyConfigCapability` as the *only*  way to modify an on-chain config if we can manage it.

Comment (runtian-zhou) [reply language/stdlib/modules/libra_configs.move]: Just fixed it!

File: language/stdlib/modules/libra_configs.move [hunk 17]
```diff
@@ -31,38 +38,113 @@ module LibraConfig {
...
+
+        move_to_sender(ModifyConfigCapability<Config> {});
+        move_to_sender(T{ payload });
+        // We don't trigger reconfiguration here, instead we'll wait for all validators update the binary
+        // to register this config into ON_CHAIN_CONFIG_REGISTRY then send another transaction to change
+        // the value which triggers the reconfiguration.
+    }
+
+    // Publish a new config item. Only the delegated address can modify such config after redeeming the capability.
+    public fun publish_new_config_with_delegate<Config: copyable>(
```
Comment (sblackshear) [inline language/stdlib/modules/libra_configs.move:121]: Is there a test that exercises this + `claim_delegated_modify_config`?

Comment (runtian-zhou) [reply language/stdlib/modules/libra_configs.move:121]: Just added the test!
